wallet/invoices.c

```diff
@@ -31,17 +39,55 @@ struct invoices {
 	struct timers *timers;
 	/* The invoice list. */
 	struct list_head invlist;
-	/* Waiters waiting for any new invoice to be paid. */
-	struct list_head waitany_waiters;
+	/* Waiters waiting for invoices to be paid, expired, or deleted. */
+	struct list_head waiters;
 };
 
 static void trigger_invoice_waiter(struct invoice_waiter *w,
-				   struct invoice *invoice)
+				   const struct invoice *invoice)
 {
 	w->triggered = true;
 	w->cb(invoice, w->cbarg);
 }
 
+static void trigger_invoice_waiter_resolve(struct invoices *invoices,
+					   u64 id,
+					   const struct invoice *invoice)
+{
+	const tal_t *tmpctx = tal_tmpctx(invoices);
+	struct invoice_waiter *w;
+	struct invoice_waiter *n;
+
+	list_for_each_safe(&invoices->waiters, w, n, list) {
+		if (!w->any && w->id != id)
+			continue;
+		list_del_from(&invoices->waiters, &w->list);
+		tal_steal(tmpctx, w);
+		trigger_invoice_waiter(w, invoice);
+	}
+
+	tal_free(tmpctx);
+}
+static void
+trigger_invoice_waiter_expire_or_delete(struct invoices *invoices,
+					u64 id,
+					const struct invoice *invoice)
+{
+	const tal_t *tmpctx = tal_tmpctx(invoices);
+	struct invoice_waiter *w;
+	struct invoice_waiter *n;
+
+	list_for_each_safe(&invoices->waiters, w, n, list) {
+		if (w->any || w->id != id)
```

Comment:
These two functions are one character different.  Perhaps we should combine them with a 'bool resolved' flag, and have a comment:

```
    if (w->any) {
            /* waitany only cares about resolved invoices */
            if (!resolved)
                    continue;
    } else if (w->id != id)
            continue;
````